File: packages/protocol/src/sample-info.ts
import { z } from "zod";
import { sampleRefSchema } from "./refs.js";
import { sampleProvenanceSchema } from "./sample-provenance.js";

/** Control-thread inspection request; no engine or project is required. */
export const inspectSampleRequestSchema = z.object({
  protocolVersion: z.string(),
  path: z.string().min(1).refine((path) => !path.includes("\0"), "path must not contain NUL"),
});
export type InspectSampleRequest = z.infer<typeof inspectSampleRequestSchema>;

/** Decoded dimensions plus original-file provenance; never contains PCM. */
export const sampleInfoSchema = sampleRefSchema.pick({
  sha256: true, format: true, sampleRate: true, channels: true, frames: true,
}).extend({
  protocolVersion: z.string(),
  sourceChannels: z.number().int().min(1).max(255),
  sourceBitDepth: z.number().int().positive().optional(),
  decoder: z.string().min(1),
  channelLayoutAction: z.enum(["kept", "downmixed-to-stereo"]),
});
export type SampleInfo = z.infer<typeof sampleInfoSchema>;

/** Explicit import-time WAV cache publication; independent of any engine. */
export const cacheSampleRequestSchema = inspectSampleRequestSchema.extend({
  cacheDir: inspectSampleRequestSchema.shape.path,
});
export type CacheSampleRequest = z.infer<typeof cacheSampleRequestSchema>;

export const cachedSampleInfoSchema = sampleInfoSchema.pick({
  protocolVersion: true, sha256: true, sampleRate: true, channels: true, frames: true,
}).extend({
  path: inspectSampleRequestSchema.shape.path,
  format: z.literal("wav"),
  provenance: sampleProvenanceSchema,
});
export type CachedSampleInfo = z.infer<typeof cachedSampleInfoSchema>;

File: crates/core/src/wire/sample_info.rs
//! Versioned control-thread sample inspection. No audio buffers cross the ABI.
use serde::{Deserialize, Serialize};

use super::SampleFormat;

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InspectSampleRequest {
    pub protocol_version: String,
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum SampleChannelLayoutAction {
    Kept,
    DownmixedToStereo,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SampleInfo {
    pub protocol_version: String,
    pub sha256: String,
    pub format: SampleFormat,
    pub sample_rate: u32,
    pub channels: u8,
    #[serde(with = "super::serde_util::u64_string")]
    pub frames: u64,
    pub source_channels: u8,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_bit_depth: Option<u8>,
    pub decoder: String,
    pub channel_layout_action: SampleChannelLayoutAction,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum SampleCacheEncoding {
    #[serde(rename = "wav-f32-v1")]
    WavF32V1,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SampleProvenance {
    pub source_sha256: String,
    pub source_format: SampleFormat,
    pub source_sample_rate: u32,
    pub source_channels: u8,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_bit_depth: Option<u8>,
    pub decoder: String,
    pub channel_layout_action: SampleChannelLayoutAction,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cache_encoding: Option<SampleCacheEncoding>,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CacheSampleRequest {
    pub protocol_version: String,
    pub path: String,
    pub cache_dir: String,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CachedSampleInfo {
    pub protocol_version: String,
    pub path: String,
    pub sha256: String,
    pub format: SampleFormat,
    pub sample_rate: u32,
    pub channels: u8,
    #[serde(with = "super::serde_util::u64_string")]
    pub frames: u64,
    pub provenance: SampleProvenance,
}
